feat: Fixed manifest-generate-paths monorepo support for Bitbucket #18968

rahulbollisetty opened the pull request:
Fixes #18968 

### PR Summary:

This PR addresses an issue with the `argocd.argoproj.io/manifest-generate-paths` [Manifest Paths Annotation](https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#manifest-paths-annotation) in Argo CD when used with Bitbucket repositories.

### Key Changes:
- Prevents spurious refreshes in cases where multiple Applications are using the same Git monorepo stored in Bitbucket.
- Ensures that the `manifest-generate-paths` annotation is handled correctly, refreshing the app only when the files changed match with the paths in  `manifest-generate-paths` .

### Steps to Implement:

1. **Add Bitbucket Repository to ArgoCD:**
   Use the following command to add the Bitbucket repository to ArgoCD with authentication:
   ```sh
   argocd repo add <REPO_URL> --username <BITBUCKET_USERNAME> --password <BITBUCKET_APP_PASSWORD>
   ```
   - ```<REPO_URL>```: Replace with your Bitbucket repository URL.
   - ```<BITBUCKET_USERNAME>```: Your Bitbucket account username.
   - ```<BITBUCKET_APP_PASSWORD>```: An app password generated in Bitbucket (can be found in your [Bitbucket account setting](https://bitbucket.org/account/settings/)

2. **Generate an Access Token for the Repository in Bitbucket:**
    - Create an access token from your Bitbucket repository.
    - Use the access token in your repository's secret.
3. **Store the Access Token as a Secret:**
Convert the generated access token to Base64 format. You can use the following command:
    ```sh
    echo "ACCESS_TOKEN" | base64 -w0
    ```
    - Replace ```ACCESS_TOKEN``` with your actual Bitbucket access token.
    - Use the access token generated in the previous step to add a Kubernetes secret for the repository added in ```Step 1```.Store the access token in the secret under the key ```data.accessToken```
    
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

andrii-korotkov-verkada commented on util/settings/settings.go:
```diff
@@ -807,6 +807,34 @@ func (mgr *SettingsManager) GetInstallationID() (string, error) {
+		if secretRepoURL == repoURL {
+			if tokenBytes, exists := secret.Data["accessToken"]; exists {
+				return string(tokenBytes), nil
+			}
```
I'm a bit uncomfortable with using selector and processing multiple items. What if somebody hijacks the flow by adding a new secret with a different/wrong token?

rahulbollisetty replied:
Could you please elaborate a bit more about this?

andrii-korotkov-verkada replied:
Why don't we just query a specific secret?

rahulbollisetty replied:
well see, if we are adding a private repo the secret that are created are of the format ```repo-178100904``` , so you cannot directly query a specific secret, you have to iterate all the repository type secret and the select the one that matches with the URL.

anandf replied:
I think we can have the same repository-access-token (RAT) used while creating the repository for doing a call from the webhook as well.

andrii-korotkov-verkada commented on util/webhook/webhook.go:
```diff
@@ -132,13 +135,24 @@ func (a *ArgoCDWebhookHandler) startWorkerPool(webhookParallelism int) {
+	// Check if a handler was passed
+	var webhookHandler *ArgoCDWebhookHandler
+	if len(optionalHandler) > 0 {
+		if handler, ok := optionalHandler[0].(*ArgoCDWebhookHandler); ok {
```
This leads to a confusing function signature, which implies multiple optional handlers can be passed. Maybe use a pointer and have it be nil if the optional handler is not available?

rahulbollisetty replied:
fixed pls check

andrii-korotkov-verkada commented on util/webhook/webhook.go:
```diff
@@ -244,14 +281,67 @@ func affectedRevisionInfo(payloadIf interface{}) (webURLs []string, revision str
+	// https://developer.atlassian.com/cloud/bitbucket/rest/api-group-commits/#api-repositories-workspace-repo-slug-diffstat-spec-get
+
+	url := fmt.Sprintf("https://api.bitbucket.org/2.0/repositories/%s/%s/diffstat/%s",
+		workspace, repoSlug, spec)
```
Fix formatting, either one argument per line or all in one line.

rahulbollisetty replied:
fixed pls check

andrii-korotkov-verkada commented on util/webhook/webhook.go:
```diff
@@ -191,9 +203,32 @@ func affectedRevisionInfo(payloadIf interface{}) (webURLs []string, revision str
+
+			if webhookHandler != nil {
+				spec := change.shaAfter + ".." + change.shaBefore
+				accessToken, err := webhookHandler.settingsSrc.GetRepositorySecretAccessToken(webURLs[0])
```
webURLs[0] doesn't read well here, i.e. what makes index 0 special?

rahulbollisetty replied:
Pls have a look at the  [webhook payload](https://support.atlassian.com/bitbucket-cloud/docs/event-payloads/?focusedCommentId=774045711#comment-774045711), you can see that the bitbucket includes multiple changes in a single event, so I have appended the URL of repo in webURL. So only one webURLs is present in case of bitbucket, whereas it is different for github, gitlab, etc. So it has to be a array, wherein only one item is present in case of bitbucket. That's why index 0 is used.

andrii-korotkov-verkada commented on util/webhook/webhook.go:
```diff
@@ -244,14 +279,66 @@ func affectedRevisionInfo(payloadIf interface{}) (webURLs []string, revision str
+		return nil, fmt.Errorf("error creating request: %w", err)
+	}
+
+	trimmedAccessToken := strings.TrimSpace(accessToken)
```
Why do we have to trim it? Why can't it be without spaces?

rahulbollisetty replied:
Whenever fetching the accessToken from the secrets, a new line is added by default. When this accessToken is passed without removing the spaces/newline, the HTTP GET call fails, with error -  ```invalid header field value for \"Authorization\""```

andrii-korotkov-verkada commented on util/webhook/webhook.go:
```diff
@@ -132,13 +135,22 @@ func (a *ArgoCDWebhookHandler) startWorkerPool(webhookParallelism int) {
 // affectedRevisionInfo examines a payload from a webhook event, and extracts the repo web URL,
 // the revision, and whether or not this affected origin/HEAD (the default branch of the repository)
-func affectedRevisionInfo(payloadIf interface{}) (webURLs []string, revision string, change changeInfo, touchedHead bool, changedFiles []string) {
+func affectedRevisionInfo(payloadIf interface{}, optionalHandler *ArgoCDWebhookHandler) (webURLs []string, revision string, change changeInfo, touchedHead bool, changedFiles []string) {
```
Why not just call this webhookHandler and use it directly without declaring extra vars?

rahulbollisetty replied:
fixed it, please check

nitishfy reviewed (commented):
I'll spend some time taking a look at the code you've written so far. However, can you briefly explain how did you solve [this](https://github.com/argoproj/argo-cd/issues/18968#issuecomment-2332115256) problem?

rahulbollisetty commented:
> I'll spend some time taking a look at the code you've written so far. Meanwhile, can you briefly explain how did you solve [this](https://github.com/argoproj/argo-cd/issues/18968#issuecomment-2332115256) problem?

The  [webhook payload](https://support.atlassian.com/bitbucket-cloud/docs/event-payloads/?focusedCommentId=774045711#comment-774045711) didn't consist of files changed. Bitbucket provides [API](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-commits/#api-repositories-workspace-repo-slug-diffstat-spec-get), which gives the path of files that are changed. Using the webhook payload, I pull out the hash of new and old commit, workspace name and repoSlug, and then get the access token from repo secret (repo secret is created when a repo is added using ```argocd repo add``` command, then user has manually add the accessToken inside the data key of repo secret), using these I make a GET call. I have added a method (```GetRepositorySecretAccessToken```) in ```util/settings.go``` to get the secret, as it was the best possible option to get the secret from the added repo when called from ```util/webhook.go```. To access the ```util/settings.go``` , I have modified the signature of ```affectedRevisionInfo``` in ```util/webhook.go``` specifically for bitbucket.

jarlebh commented:
We really need this fix, any timeline for getting this merged ?

nitishfy commented:
I'll start with reviewing this PR in depth now.

nitishfy commented on util/webhook/webhook.go:
```diff
@@ -191,9 +197,32 @@ func affectedRevisionInfo(payloadIf interface{}) (webURLs []string, revision str
 		webURLs = append(webURLs, payload.Repository.Links.HTML.Href)
 		// TODO: bitbucket includes multiple changes as part of a single event.
 		// We only pick the first but need to consider how to handle multiple
```
```suggestion
```

nitishfy reviewed (approved):
Tested and works well. We need to update the docs to tell about the implementation which can be done in the same PR.

nitishfy commented on util/webhook/webhook.go:
```diff
@@ -244,14 +273,66 @@ func affectedRevisionInfo(payloadIf interface{}) (webURLs []string, revision str
 }
 
+func fetchDiffstatFromBitbucket(workspace, repoSlug, spec, accessToken string) ([]string, error) {
+	// Getting the files changed from diff API:
+	// https://developer.atlassian.com/cloud/bitbucket/rest/api-group-commits/#api-repositories-workspace-repo-slug-diffstat-spec-get
```
```suggestion
// fetchDiffstatFromBitbucket gets the files changed from diff API - https://developer.atlassian.com/cloud/bitbucket/rest/api-group-commits/#api-repositories-workspace-repo-slug-diffstat-spec-get
```
Please add this as a function comment (at the top of function)

crenshaw-dev commented on util/webhook/webhook.go:
```diff
@@ -244,14 +271,66 @@ func affectedRevisionInfo(payloadIf interface{}) (webURLs []string, revision str
 	return webURLs, revision, change, touchedHead, changedFiles
 }
 
+func fetchDiffstatFromBitbucket(workspace, repoSlug, spec, accessToken string) ([]string, error) {
```
I'm concerned about this. iirc our webhook handler isn't auth-protected. So anyone with API access could hammer the endpoint and trigger SSRF. I don't have time to give a thorough review right now, but requesting changes so this doesn't get merged without further conversation.

anandf commented:
I have created an improved version of this PR. https://github.com/argoproj/argo-cd/pull/21811/files
the improvements I have done in my branch are,
1. Use existing repo secret for the webhook callback whereas the original one required a new kind of secret to store auth tokens for making calls from within the webhook handler.
2. Use BitBucket APIs to do the REST call instead of using http client.

nitishfy commented:
Closing this in favour of https://github.com/argoproj/argo-cd/pull/21811
